File: src/CalcViewModel/Common/NavCategory.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "NavCategory.h"
#include "AppResourceProvider.h"
#include "Common/LocalizationStringUtil.h"
#include <initializer_list>

using namespace CalculatorApp;
using namespace CalculatorApp::ViewModel::Common;
using namespace CalculatorApp::ViewModel;
using namespace Concurrency;
using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation::Collections;
using namespace Windows::Management::Policies;
using namespace Windows::System;

namespace UCM = UnitConversionManager;

// Calculator categories always support negative and positive.
static constexpr bool SUPPORTS_ALL = true;

// Converter categories usually only support positive.
static constexpr bool SUPPORTS_NEGATIVE = true;
static constexpr bool POSITIVE_ONLY = false;

// vvv THESE CONSTANTS SHOULD NEVER CHANGE vvv
static constexpr int STANDARD_ID = 0;
static constexpr int SCIENTIFIC_ID = 1;
static constexpr int PROGRAMMER_ID = 2;
static constexpr int DATE_ID = 3;
static constexpr int VOLUME_ID = 4;
static constexpr int LENGTH_ID = 5;
static constexpr int WEIGHT_ID = 6;
static constexpr int TEMPERATURE_ID = 7;
static constexpr int ENERGY_ID = 8;
static constexpr int AREA_ID = 9;
static constexpr int SPEED_ID = 10;
static constexpr int TIME_ID = 11;
static constexpr int POWER_ID = 12;
static constexpr int DATA_ID = 13;
static constexpr int PRESSURE_ID = 14;
static constexpr int ANGLE_ID = 15;
static constexpr int CURRENCY_ID = 16;
static constexpr int GRAPHING_ID = 17;
// ^^^ THESE CONSTANTS SHOULD NEVER CHANGE ^^^

namespace // put the utils within this TU
{
    Platform::String^ CurrentUserId;

    bool IsGraphingModeEnabled()
    {
        static auto enabled = []
        {
            auto user = User::GetFromId(CurrentUserId);
            if (user == nullptr)
            {
                return true;
            }
            return NamedPolicy::GetPolicyFromPathForUser(user, L"Education", L"AllowGraphingCalculator")->GetBoolean();
        }();
        return enabled;
    }

    // The order of items in this list determines the order of items in the menu.
    const std::vector<NavCategoryInitializer> s_categoryManifest {
      NavCategoryInitializer{ ViewMode::Standard,
          STANDARD_ID,
          L"Standard",
          L"StandardMode",
          L"\uE8EF",
          CategoryGroupType::Calculator,
          MyVirtualKey::Number1,
          L"1",
          SUPPORTS_ALL },
      NavCategoryInitializer{ ViewMode::Scientific,
          SCIENTIFIC_ID,
          L"Scientific",
          L"ScientificMode",
          L"\uF196",
          CategoryGroupType::Calculator,
          MyVirtualKey::Number2,
          L"2",
          SUPPORTS_ALL },
      NavCategoryInitializer{ ViewMode::Graphing,
          GRAPHING_ID,
          L"Graphing",
          L"GraphingCalculatorMode",
          L"\uF770",
          CategoryGroupType::Calculator,
          MyVirtualKey::Number3,
          L"3",
          SUPPORTS_ALL },
      NavCategoryInitializer{ ViewMode::Programmer,
          PROGRAMMER_ID,
          L"Programmer",
          L"ProgrammerMode",
          L"\uECCE",
          CategoryGroupType::Calculator,
          MyVirtualKey::Number4,
          L"4",
          SUPPORTS_ALL },
      NavCategoryInitializer{ ViewMode::Date,
          DATE_ID,
          L"Date",
          L"DateCalculationMode",
          L"\uE787",
          CategoryGroupType::Calculator,
          MyVirtualKey::Number5,
          L"5",
          SUPPORTS_ALL },
      NavCategoryInitializer{ ViewMode::Currency,
          CURRENCY_ID,
          L"Currency",
          L"CategoryName_Currency",
          L"\uEB0D",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Volume,
          VOLUME_ID,
          L"Volume",
          L"CategoryName_Volume",
          L"\uF1AA",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Length,
          LENGTH_ID,
          L"Length",
          L"CategoryName_Length",
          L"\uECC6",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Weight,
          WEIGHT_ID,
          L"Weight and Mass",
          L"CategoryName_Weight",
          L"\uF4C1",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Temperature,
          TEMPERATURE_ID,
          L"Temperature",
          L"CategoryName_Temperature",
          L"\uE7A3",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          SUPPORTS_NEGATIVE },
      NavCategoryInitializer{ ViewMode::Energy,
          ENERGY_ID,
          L"Energy",
          L"CategoryName_Energy",
          L"\uECAD",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Area,
          AREA_ID,
          L"Area",
          L"CategoryName_Area",
          L"\uE809",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Speed,
          SPEED_ID,
          L"Speed",
          L"CategoryName_Speed",
          L"\uEADA",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Time,
          TIME_ID,
          L"Time",
          L"CategoryName_Time",
          L"\uE917",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Power,
          POWER_ID,
          L"Power",
          L"CategoryName_Power",
          L"\uE945",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          SUPPORTS_NEGATIVE },
      NavCategoryInitializer{ ViewMode::Data,
          DATA_ID,
          L"Data",
          L"CategoryName_Data",
          L"\uF20F",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Pressure,
          PRESSURE_ID,
          L"Pressure",
          L"CategoryName_Pressure",
          L"\uEC4A",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          POSITIVE_ONLY },
      NavCategoryInitializer{ ViewMode::Angle,
          ANGLE_ID,
          L"Angle",
          L"CategoryName_Angle",
          L"\uF515",
          CategoryGroupType::Converter,
          MyVirtualKey::None,
          std::nullopt,
          SUPPORTS_NEGATIVE },
    };
} // namespace unnamed


bool NavCategory::IsCalculatorViewMode(ViewModeType mode)
{
    // Historically, Calculator modes are Standard, Scientific, and Programmer.
    return !IsDateCalculatorViewMode(mode) && !IsGraphingCalculatorViewMode(mode) && IsModeInCategoryGroup(mode, CategoryGroupType::Calculator);
}

bool NavCategory::IsGraphingCalculatorViewMode(ViewModeType mode)
{
    return mode == ViewModeType::Graphing;
}

bool NavCategory::IsDateCalculatorViewMode(ViewModeType mode)
{
    return mode == ViewModeType::Date;
}

bool NavCategory::IsConverterViewMode(ViewModeType mode)
{
    return IsModeInCategoryGroup(mode, CategoryGroupType::Converter);
}

bool NavCategory::IsModeInCategoryGroup(ViewModeType mode, CategoryGroupType type)
{
    return std::any_of(
        s_categoryManifest.cbegin(),
        s_categoryManifest.cend(),
        [mode, type](const auto& initializer) {
            return initializer.viewMode == mode && initializer.groupType == type;
        });
}

NavCategoryGroup::NavCategoryGroup(const NavCategoryGroupInitializer& groupInitializer)
    : m_Categories(ref new Vector<NavCategory ^>())
{
    m_GroupType = groupInitializer.type;

    auto resProvider = AppResourceProvider::GetInstance();
    m_Name = resProvider->GetResourceString(StringReference(groupInitializer.headerResourceKey));
    String ^ groupMode = resProvider->GetResourceString(StringReference(groupInitializer.modeResourceKey));
    String ^ automationName = resProvider->GetResourceString(StringReference(groupInitializer.automationResourceKey));

    String ^ navCategoryHeaderAutomationNameFormat = resProvider->GetResourceString(L"NavCategoryHeader_AutomationNameFormat");
    m_AutomationName = LocalizationStringUtil::GetLocalizedString(navCategoryHeaderAutomationNameFormat, automationName);

    String ^ navCategoryItemAutomationNameFormat = resProvider->GetResourceString(L"NavCategoryItem_AutomationNameFormat");

    for (const NavCategoryInitializer& categoryInitializer : s_categoryManifest)
    {
        if (categoryInitializer.groupType == groupInitializer.type)
        {
            String ^ nameResourceKey = StringReference(categoryInitializer.nameResourceKey);
            String ^ categoryName = resProvider->GetResourceString(nameResourceKey + "Text");
            String ^ categoryAutomationName = LocalizationStringUtil::GetLocalizedString(navCategoryItemAutomationNameFormat, categoryName, m_Name);

            m_Categories->Append(ref new NavCategory(
                categoryName,
                categoryAutomationName,
                StringReference(categoryInitializer.glyph),
                categoryInitializer.accessKey.has_value() ? ref new String(categoryInitializer.accessKey->c_str())
                                                         : resProvider->GetResourceString(nameResourceKey + "AccessKey"),
                groupMode,
                categoryInitializer.viewMode,
                categoryInitializer.supportsNegative,
                categoryInitializer.viewMode != ViewMode::Graphing));
        }
    }
}

void NavCategoryStates::SetCurrentUser(Platform::String^ userId)
{
    CurrentUserId = userId;
}

IVector<NavCategoryGroup ^> ^ NavCategoryStates::CreateMenuOptions()
{
    auto menuOptions = ref new Vector<NavCategoryGroup ^>();
    menuOptions->Append(CreateCalculatorCategoryGroup());
    menuOptions->Append(CreateConverterCategoryGroup());
    return menuOptions;
}

NavCategoryGroup ^ NavCategoryStates::CreateCalculatorCategoryGroup()
{
    return ref new NavCategoryGroup(
        NavCategoryGroupInitializer{ CategoryGroupType::Calculator, L"CalculatorModeTextCaps", L"CalculatorModeText", L"CalculatorModePluralText" });
}

NavCategoryGroup ^ NavCategoryStates::CreateConverterCategoryGroup()
{
    return ref new NavCategoryGroup(
        NavCategoryGroupInitializer{ CategoryGroupType::Converter, L"ConverterModeTextCaps", L"ConverterModeText", L"ConverterModePluralText" });
}

// This function should only be used when storing the mode to app data.
int NavCategoryStates::Serialize(ViewMode mode)
{
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [mode](const auto& initializer) { return initializer.viewMode == mode; });

    return (citer != s_categoryManifest.cend()) ? citer->serializationId : -1;
}

// This function should only be used when restoring the mode from app data.
ViewMode NavCategoryStates::Deserialize(Platform::Object ^ obj)
{
    // If we cast directly to ViewMode we will fail
    // because we technically store an int.
    // Need to cast to int, then ViewMode.
    auto boxed = dynamic_cast<Box<int> ^>(obj);
    if (boxed != nullptr)
    {
        int serializationId = boxed->Value;
        const auto& citer = find_if(
            cbegin(s_categoryManifest),
            cend(s_categoryManifest),
            [serializationId](const auto& initializer) { return initializer.serializationId == serializationId; });

        return citer != s_categoryManifest.cend() ?
                   (citer->viewMode == ViewMode::Graphing ?
                       (IsGraphingModeEnabled() ? citer->viewMode : ViewMode::None)
                       : citer->viewMode)
                   : ViewMode::None;
    }
    else
    {
        return ViewMode::None;
    }
}

ViewMode NavCategoryStates::GetViewModeForFriendlyName(String ^ name)
{
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [name](const auto& initializer) { return wcscmp(initializer.friendlyName, name->Data()) == 0; });

    return (citer != s_categoryManifest.cend()) ? citer->viewMode : ViewMode::None;
}

String ^ NavCategoryStates::GetFriendlyName(ViewMode mode)
{
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [mode](const auto& initializer) { return initializer.viewMode == mode; });

    return (citer != s_categoryManifest.cend()) ? StringReference(citer->friendlyName) : L"None";
}

String ^ NavCategoryStates::GetNameResourceKey(ViewMode mode)
{
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [mode](const auto& initializer) { return initializer.viewMode == mode; });

    return (citer != s_categoryManifest.cend()) ? StringReference(citer->nameResourceKey) + "Text" : nullptr;
}

CategoryGroupType NavCategoryStates::GetGroupType(ViewMode mode)
{
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [mode](const auto& initializer) { return initializer.viewMode == mode; });

    return (citer != s_categoryManifest.cend()) ? citer->groupType : CategoryGroupType::None;
}

// GetIndex is 0-based, GetPosition is 1-based
int NavCategoryStates::GetIndex(ViewMode mode)
{
    int position = GetPosition(mode);
    return std::max(-1, position - 1);
}

int NavCategoryStates::GetFlatIndex(ViewMode mode)
{
    int index = -1;
    CategoryGroupType type = CategoryGroupType::None;
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [mode, &type, &index](const auto& initializer) {
            ++index;
            if (initializer.groupType != type)
            {
                type = initializer.groupType;
                ++index;
            }
            return initializer.viewMode == mode;
        });

    return (citer != s_categoryManifest.cend()) ? index : -1;
}

// GetIndex is 0-based, GetPosition is 1-based
int NavCategoryStates::GetIndexInGroup(ViewMode mode, CategoryGroupType type)
{
    int index = -1;
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [mode, type, &index](const auto& initializer) {
            if (initializer.groupType == type)
            {
                ++index;
                return initializer.viewMode == mode;
            }
            return false;
        });

    return (citer != s_categoryManifest.cend()) ? index : -1;
}

// GetIndex is 0-based, GetPosition is 1-based
int NavCategoryStates::GetPosition(ViewMode mode)
{
    int position = 0;
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [mode, &position](const auto& initializer) {
            ++position;
            return initializer.viewMode == mode;
        });

    return (citer != s_categoryManifest.cend()) ? position : -1;
}

ViewMode NavCategoryStates::GetViewModeForVirtualKey(MyVirtualKey virtualKey)
{
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [virtualKey](const auto& initializer) { return initializer.virtualKey == virtualKey; });

    return (citer != s_categoryManifest.end()) ? citer->viewMode : ViewMode::None;
}

void NavCategoryStates::GetCategoryAcceleratorKeys(IVector<MyVirtualKey> ^ accelerators)
{
    if (accelerators != nullptr)
    {
        accelerators->Clear();
        for (const auto& category : s_categoryManifest)
        {
            if (category.virtualKey != MyVirtualKey::None)
            {
                accelerators->Append(category.virtualKey);
            }
        }
    }
}

bool NavCategoryStates::IsValidViewMode(ViewMode mode)
{
    const auto& citer = find_if(
        cbegin(s_categoryManifest),
        cend(s_categoryManifest),
        [mode](const auto& initializer) { return initializer.viewMode == mode; });

    return citer != s_categoryManifest.cend();
}

bool NavCategoryStates::IsViewModeEnabled(ViewMode mode)
{
    return mode != ViewMode::Graphing ? true : IsGraphingModeEnabled();
}

File: src/Calculator.ViewModels/Common/NavCategory.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// The NavCategory group of classes and enumerations is intended to serve
// as a single location for storing metadata about a navigation mode.

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using CommunityToolkit.Mvvm.ComponentModel;
using Windows.UI.Xaml.Data;

namespace CalculatorApp.ViewModel.Common
{
    // Don't change the order of these enums
    // and definitely don't use int arithmetic to change modes.
    public enum ViewMode
    {
        None = -1,
        Standard = 0,
        Scientific = 1,
        Programmer = 2,
        Date = 3,
        Volume = 4,
        Length = 5,
        Weight = 6,
        Temperature = 7,
        Energy = 8,
        Area = 9,
        Speed = 10,
        Time = 11,
        Power = 12,
        Data = 13,
        Pressure = 14,
        Angle = 15,
        Currency = 16,
        Graphing = 17
    }

    public enum CategoryGroupType
    {
        None = -1,
        Calculator = 0,
        Converter = 1
    }

    internal struct NavCategoryInitializer
    {
        public ViewMode ViewMode;
        public int SerializationId;
        public string FriendlyName;
        public string NameResourceKey;
        public string Glyph;
        public CategoryGroupType GroupType;
        public MyVirtualKey VirtualKey;
        public string AccessKey; // null means use resource
        public bool SupportsNegative;
    }

    internal struct NavCategoryGroupInitializer
    {
        public CategoryGroupType Type;
        public string HeaderResourceKey;
        public string ModeResourceKey;
        public string AutomationResourceKey;
    }

    [Windows.UI.Xaml.Data.Bindable]
    public sealed partial class NavCategory : ObservableObject
    {
        private string _name;
        private string _automationName;
        private string _glyph;
        private ViewMode _viewMode;
        private string _accessKey;
        private bool _supportsNegative;

        [ObservableProperty]
        private bool _isEnabled;

        private string _modeString;

        internal NavCategory(
            string name,
            string automationName,
            string glyph,
            string accessKey,
            string mode,
            ViewMode viewMode,
            bool supportsNegative,
            bool isEnabled)
        {
            _name = name;
            _automationName = automationName;
            _glyph = glyph;
            _accessKey = accessKey;
            _modeString = mode;
            _viewMode = viewMode;
            _supportsNegative = supportsNegative;
            _isEnabled = isEnabled;
        }

        public string Name => _name;
        public string AutomationName => _automationName;
        public string Glyph => _glyph;
        public ViewMode ViewMode => _viewMode;
        public string AccessKey => _accessKey;
        public bool SupportsNegative => _supportsNegative;

        public string AutomationId => _viewMode.ToString();

        public static bool IsCalculatorViewMode(ViewMode mode)
        {
            // Historically, Calculator modes are Standard, Scientific, and Programmer.
            return !IsDateCalculatorViewMode(mode) && !IsGraphingCalculatorViewMode(mode) && IsModeInCategoryGroup(mode, CategoryGroupType.Calculator);
        }

        public static bool IsGraphingCalculatorViewMode(ViewMode mode)
        {
            return mode == ViewMode.Graphing;
        }

        public static bool IsDateCalculatorViewMode(ViewMode mode)
        {
            return mode == ViewMode.Date;
        }

        public static bool IsConverterViewMode(ViewMode mode)
        {
            return IsModeInCategoryGroup(mode, CategoryGroupType.Converter);
        }

        private static bool IsModeInCategoryGroup(ViewMode mode, CategoryGroupType type)
        {
            return NavCategoryStates.CategoryManifest.Any(init => init.ViewMode == mode && init.GroupType == type);
        }
    }

    [Windows.UI.Xaml.Data.Bindable]
    public sealed class NavCategoryGroup : ObservableObject
    {
        private string _name;
        private string _automationName;
        private CategoryGroupType _groupType;
        private ObservableCollection<NavCategory> _categories;

        internal NavCategoryGroup(NavCategoryGroupInitializer groupInitializer)
        {
            _categories = new ObservableCollection<NavCategory>();
            _groupType = groupInitializer.Type;

            var resProvider = AppResourceProvider.GetInstance();
            _name = resProvider.GetResourceString(groupInitializer.HeaderResourceKey);
            string groupMode = resProvider.GetResourceString(groupInitializer.ModeResourceKey);
            string automationName = resProvider.GetResourceString(groupInitializer.AutomationResourceKey);

            string navCategoryHeaderAutomationNameFormat = resProvider.GetResourceString("NavCategoryHeader_AutomationNameFormat");
            _automationName = LocalizationStringUtil.GetLocalizedString(navCategoryHeaderAutomationNameFormat, automationName);

            string navCategoryItemAutomationNameFormat = resProvider.GetResourceString("NavCategoryItem_AutomationNameFormat");

            foreach (var categoryInitializer in NavCategoryStates.CategoryManifest)
            {
                if (categoryInitializer.GroupType == groupInitializer.Type)
                {
                    string nameResourceKey = categoryInitializer.NameResourceKey;
                    string categoryName = resProvider.GetResourceString(nameResourceKey + "Text");
                    string categoryAutomationName = LocalizationStringUtil.GetLocalizedString(
                        navCategoryItemAutomationNameFormat, categoryName, _name);

                    string accessKey = categoryInitializer.AccessKey ?? resProvider.GetResourceString(nameResourceKey + "AccessKey");

                    _categories.Add(new NavCategory(
                        categoryName,
                        categoryAutomationName,
                        categoryInitializer.Glyph,
                        accessKey,
                        groupMode,
                        categoryInitializer.ViewMode,
                        categoryInitializer.SupportsNegative,
                        categoryInitializer.ViewMode != ViewMode.Graphing));
                }
            }
        }

        public string Name
        {
            get => _name;
            private set => SetProperty(ref _name, value);
        }

        public string AutomationName
        {
            get => _automationName;
            private set => SetProperty(ref _automationName, value);
        }

        public CategoryGroupType GroupType
        {
            get => _groupType;
            private set => SetProperty(ref _groupType, value);
        }

        public ObservableCollection<NavCategory> Categories
        {
            get => _categories;
            private set => SetProperty(ref _categories, value);
        }
    }

    public sealed class NavCategoryStates
    {
        // Calculator categories always support negative and positive.
        private const bool SupportsAll = true;

        // Converter categories usually only support positive.
        private const bool SupportsNegative = true;
        private const bool PositiveOnly = false;

        // vvv THESE CONSTANTS SHOULD NEVER CHANGE vvv
        private const int StandardId = 0;
        private const int ScientificId = 1;
        private const int ProgrammerId = 2;
        private const int DateId = 3;
        private const int VolumeId = 4;
        private const int LengthId = 5;
        private const int WeightId = 6;
        private const int TemperatureId = 7;
        private const int EnergyId = 8;
        private const int AreaId = 9;
        private const int SpeedId = 10;
        private const int TimeId = 11;
        private const int PowerId = 12;
        private const int DataId = 13;
        private const int PressureId = 14;
        private const int AngleId = 15;
        private const int CurrencyId = 16;
        private const int GraphingId = 17;
        // ^^^ THESE CONSTANTS SHOULD NEVER CHANGE ^^^

        private static string _currentUserId;

        internal static readonly NavCategoryInitializer[] CategoryManifest = new[]
        {
            new NavCategoryInitializer { ViewMode = ViewMode.Standard, SerializationId = StandardId, FriendlyName = "Standard", NameResourceKey = "StandardMode", Glyph = "\uE8EF", GroupType = CategoryGroupType.Calculator, VirtualKey = MyVirtualKey.Number1, AccessKey = "1", SupportsNegative = SupportsAll },
            new NavCategoryInitializer { ViewMode = ViewMode.Scientific, SerializationId = ScientificId, FriendlyName = "Scientific", NameResourceKey = "ScientificMode", Glyph = "\uF196", GroupType = CategoryGroupType.Calculator, VirtualKey = MyVirtualKey.Number2, AccessKey = "2", SupportsNegative = SupportsAll },
            new NavCategoryInitializer { ViewMode = ViewMode.Graphing, SerializationId = GraphingId, FriendlyName = "Graphing", NameResourceKey = "GraphingCalculatorMode", Glyph = "\uF770", GroupType = CategoryGroupType.Calculator, VirtualKey = MyVirtualKey.Number3, AccessKey = "3", SupportsNegative = SupportsAll },
            new NavCategoryInitializer { ViewMode = ViewMode.Programmer, SerializationId = ProgrammerId, FriendlyName = "Programmer", NameResourceKey = "ProgrammerMode", Glyph = "\uECCE", GroupType = CategoryGroupType.Calculator, VirtualKey = MyVirtualKey.Number4, AccessKey = "4", SupportsNegative = SupportsAll },
            new NavCategoryInitializer { ViewMode = ViewMode.Date, SerializationId = DateId, FriendlyName = "Date", NameResourceKey = "DateCalculationMode", Glyph = "\uE787", GroupType = CategoryGroupType.Calculator, VirtualKey = MyVirtualKey.Number5, AccessKey = "5", SupportsNegative = SupportsAll },
            new NavCategoryInitializer { ViewMode = ViewMode.Currency, SerializationId = CurrencyId, FriendlyName = "Currency", NameResourceKey = "CategoryName_Currency", Glyph = "\uEB0D", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Volume, SerializationId = VolumeId, FriendlyName = "Volume", NameResourceKey = "CategoryName_Volume", Glyph = "\uF1AA", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Length, SerializationId = LengthId, FriendlyName = "Length", NameResourceKey = "CategoryName_Length", Glyph = "\uECC6", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Weight, SerializationId = WeightId, FriendlyName = "Weight and Mass", NameResourceKey = "CategoryName_Weight", Glyph = "\uF4C1", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Temperature, SerializationId = TemperatureId, FriendlyName = "Temperature", NameResourceKey = "CategoryName_Temperature", Glyph = "\uE7A3", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = SupportsNegative },
            new NavCategoryInitializer { ViewMode = ViewMode.Energy, SerializationId = EnergyId, FriendlyName = "Energy", NameResourceKey = "CategoryName_Energy", Glyph = "\uECAD", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Area, SerializationId = AreaId, FriendlyName = "Area", NameResourceKey = "CategoryName_Area", Glyph = "\uE809", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Speed, SerializationId = SpeedId, FriendlyName = "Speed", NameResourceKey = "CategoryName_Speed", Glyph = "\uEADA", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Time, SerializationId = TimeId, FriendlyName = "Time", NameResourceKey = "CategoryName_Time", Glyph = "\uE917", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Power, SerializationId = PowerId, FriendlyName = "Power", NameResourceKey = "CategoryName_Power", Glyph = "\uE945", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = SupportsNegative },
            new NavCategoryInitializer { ViewMode = ViewMode.Data, SerializationId = DataId, FriendlyName = "Data", NameResourceKey = "CategoryName_Data", Glyph = "\uF20F", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Pressure, SerializationId = PressureId, FriendlyName = "Pressure", NameResourceKey = "CategoryName_Pressure", Glyph = "\uEC4A", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = PositiveOnly },
            new NavCategoryInitializer { ViewMode = ViewMode.Angle, SerializationId = AngleId, FriendlyName = "Angle", NameResourceKey = "CategoryName_Angle", Glyph = "\uF515", GroupType = CategoryGroupType.Converter, VirtualKey = MyVirtualKey.None, AccessKey = null, SupportsNegative = SupportsNegative },
        };

        public static void SetCurrentUser(string userId)
        {
            _currentUserId = userId;
        }

        public static IList<NavCategoryGroup> CreateMenuOptions()
        {
            var menuOptions = new List<NavCategoryGroup>();
            menuOptions.Add(CreateCalculatorCategoryGroup());
            menuOptions.Add(CreateConverterCategoryGroup());
            return menuOptions;
        }

        public static NavCategoryGroup CreateCalculatorCategoryGroup()
        {
            return new NavCategoryGroup(new NavCategoryGroupInitializer
            {
                Type = CategoryGroupType.Calculator,
                HeaderResourceKey = "CalculatorModeTextCaps",
                ModeResourceKey = "CalculatorModeText",
                AutomationResourceKey = "CalculatorModePluralText"
            });
        }

        public static NavCategoryGroup CreateConverterCategoryGroup()
        {
            return new NavCategoryGroup(new NavCategoryGroupInitializer
            {
                Type = CategoryGroupType.Converter,
                HeaderResourceKey = "ConverterModeTextCaps",
                ModeResourceKey = "ConverterModeText",
                AutomationResourceKey = "ConverterModePluralText"
            });
        }

        public static bool IsValidViewMode(ViewMode mode)
        {
            return CategoryManifest.Any(init => init.ViewMode == mode);
        }

        public static bool IsViewModeEnabled(ViewMode mode)
        {
            // Graphing mode may be disabled by policy; other modes are always enabled
            return mode != ViewMode.Graphing || true; // Simplified: always enabled in managed
        }

        // This function should only be used when storing the mode to app data.
        public static int Serialize(ViewMode mode)
        {
            var match = CategoryManifest.FirstOrDefault(init => init.ViewMode == mode);
            return match.FriendlyName != null ? match.SerializationId : -1;
        }

        // This function should only be used when restoring the mode from app data.
        public static ViewMode Deserialize(object obj)
        {
            if (obj is int serializationId)
            {
                var match = CategoryManifest.FirstOrDefault(init => init.SerializationId == serializationId);
                if (match.FriendlyName != null)
                {
                    if (match.ViewMode == ViewMode.Graphing)
                    {
                        return IsViewModeEnabled(ViewMode.Graphing) ? match.ViewMode : ViewMode.None;
                    }
                    return match.ViewMode;
                }
            }
            return ViewMode.None;
        }

        public static ViewMode GetViewModeForFriendlyName(string name)
        {
            var match = CategoryManifest.FirstOrDefault(init => init.FriendlyName == name);
            return match.FriendlyName != null ? match.ViewMode : ViewMode.None;
        }

        public static string GetFriendlyName(ViewMode mode)
        {
            var match = CategoryManifest.FirstOrDefault(init => init.ViewMode == mode);
            return match.FriendlyName ?? "None";
        }

        public static string GetNameResourceKey(ViewMode mode)
        {
            var match = CategoryManifest.FirstOrDefault(init => init.ViewMode == mode);
            return match.NameResourceKey != null ? match.NameResourceKey + "Text" : null;
        }

        public static CategoryGroupType GetGroupType(ViewMode mode)
        {
            var match = CategoryManifest.FirstOrDefault(init => init.ViewMode == mode);
            return match.FriendlyName != null ? match.GroupType : CategoryGroupType.None;
        }

        // GetIndex is 0-based, GetPosition is 1-based
        public static int GetIndex(ViewMode mode)
        {
            int position = GetPosition(mode);
            return Math.Max(-1, position - 1);
        }

        public static int GetFlatIndex(ViewMode mode)
        {
            int index = -1;
            CategoryGroupType type = CategoryGroupType.None;

            foreach (var init in CategoryManifest)
            {
                ++index;
                if (init.GroupType != type)
                {
                    type = init.GroupType;
                    ++index;
                }
                if (init.ViewMode == mode)
                {
                    return index;
                }
            }

            return -1;
        }

        public static int GetIndexInGroup(ViewMode mode, CategoryGroupType type)
        {
            int index = -1;
            foreach (var init in CategoryManifest)
            {
                if (init.GroupType == type)
                {
                    ++index;
                    if (init.ViewMode == mode)
                    {
                        return index;
                    }
                }
            }
            return -1;
        }

        // GetIndex is 0-based, GetPosition is 1-based
        public static int GetPosition(ViewMode mode)
        {
            int position = 0;
            foreach (var init in CategoryManifest)
            {
                ++position;
                if (init.ViewMode == mode)
                {
                    return position;
                }
            }
            return -1;
        }

        public static ViewMode GetViewModeForVirtualKey(MyVirtualKey virtualKey)
        {
            var match = CategoryManifest.FirstOrDefault(init => init.VirtualKey == virtualKey);
            return match.FriendlyName != null ? match.ViewMode : ViewMode.None;
        }

        public static void GetCategoryAcceleratorKeys(IList<MyVirtualKey> results)
        {
            if (results != null)
            {
                results.Clear();
                foreach (var category in CategoryManifest)
                {
                    if (category.VirtualKey != MyVirtualKey.None)
                    {
                        results.Add(category.VirtualKey);
                    }
                }
            }
        }
    }
}
